File: selfdrive/ui/qt/widgets/prime.cc
#include "selfdrive/ui/qt/widgets/prime.h"

#include <QLabel>
#include <QStackedWidget>
#include <QVBoxLayout>

PrimeUserWidget::PrimeUserWidget(QWidget* parent) : QFrame(parent) {
  setObjectName("primeWidget");
  QVBoxLayout *mainLayout = new QVBoxLayout(this);
  mainLayout->setContentsMargins(30, 20, 30, 20);
  mainLayout->setSpacing(20);

  QLabel *subscribed = new QLabel(tr("✓ OFFLINE"));
  subscribed->setStyleSheet("font-size: 41px; font-weight: bold; color: #86FF4E;");
  mainLayout->addWidget(subscribed);

  QLabel *eopLabel = new QLabel(tr("EnhancedOpenPilot"));
  eopLabel->setStyleSheet("font-size: 45px; font-weight: bold;");
  mainLayout->addWidget(eopLabel);

  QLabel *desc = new QLabel(tr("All processing is local. No cloud account required."));
  desc->setStyleSheet("font-size: 45px; font-weight: light; color: #CCCCCC;");
  desc->setWordWrap(true);
  mainLayout->addWidget(desc);

  setStyleSheet(R"(
    PrimeUserWidget {
      border-radius: 10px;
      background-color: #333333;
    }
  )");
}

PrimeAdWidget::PrimeAdWidget(QWidget* parent) : QFrame(parent) {
  QVBoxLayout *main_layout = new QVBoxLayout(this);
  main_layout->setContentsMargins(30, 20, 30, 20);
  main_layout->setSpacing(20);

  QLabel *title = new QLabel(tr("EnhancedOpenPilot"));
  title->setStyleSheet("font-size: 45px; font-weight: bold;");
  main_layout->addWidget(title, 0, Qt::AlignTop);

  QLabel *desc = new QLabel(tr("EOP is ready to use. No cloud account or subscription required."));
  desc->setStyleSheet("font-size: 30px; font-weight: light; color: white;");
  desc->setWordWrap(true);
  main_layout->addWidget(desc, 0, Qt::AlignTop);

  main_layout->addStretch();

  setStyleSheet(R"(
    PrimeAdWidget {
      border-radius: 10px;
      background-color: #333333;
    }
  )");
}

SetupWidget::SetupWidget(QWidget* parent) : QFrame(parent) {
  mainLayout = new QStackedWidget;

  // EOP setup prompt (no pairing)
  QFrame* eopSetup = new QFrame;
  eopSetup->setObjectName("primeWidget");
  QVBoxLayout* setupLayout = new QVBoxLayout(eopSetup);
  setupLayout->setSpacing(38);
  setupLayout->setContentsMargins(30, 20, 30, 20);

  QLabel* title = new QLabel(tr("EOP Setup"));
  title->setStyleSheet("font-size: 45px; font-weight: bold;");
  setupLayout->addWidget(title);

  QLabel* description = new QLabel(tr("EnhancedOpenPilot (EOP) is ready to use. No cloud account required."));
  description->setWordWrap(true);
  description->setStyleSheet("font-size: 30px; font-weight: light;");
  setupLayout->addWidget(description);

  QLabel* desc2 = new QLabel(tr("All AI processing is local via Hailo-8. No data leaves your device."));
  desc2->setWordWrap(true);
  desc2->setStyleSheet("font-size: 40px; font-weight: light; color: #AAAAAA;");
  setupLayout->addWidget(desc2);

  setupLayout->addStretch();

  mainLayout->addWidget(eopSetup);

  QVBoxLayout *outer_layout = new QVBoxLayout(this);
  outer_layout->setContentsMargins(0, 0, 0, 0);
  outer_layout->addWidget(mainLayout);

  setStyleSheet(R"(
    #primeWidget {
      border-radius: 10px;
      background-color: #333333;
    }
  )");

  // Retain size while hidden
  QSizePolicy sp_retain = sizePolicy();
  sp_retain.setRetainSizeWhenHidden(true);
  setSizePolicy(sp_retain);
}

File: selfdrive/ui/widgets/prime.py
import pyray as rl

from openpilot.system.ui.lib.application import gui_app, FontWeight
from openpilot.system.ui.lib.wrap_text import wrap_text
from openpilot.system.ui.widgets import Widget


class PrimeWidget(Widget):
  """Widget for displaying EOP status — no cloud subscription."""

  BG_COLOR = rl.Color(51, 51, 51, 255)

  def _render(self, rect):
    self._render_eop_status(rect)

  def _render_eop_status(self, rect: rl.Rectangle):
    """Renders the EOP offline status widget."""

    rl.draw_rectangle_rounded(rect, 0.02, 10, self.BG_COLOR)

    x = rect.x + 56
    y = rect.y + 40
    w = rect.width - 112

    font = gui_app.font(FontWeight.BOLD)
    rl.draw_text_ex(font, "✓ OFFLINE", rl.Vector2(x, y), 41, 0, rl.Color(134, 255, 78, 255))
    rl.draw_text_ex(font, "EnhancedOpenPilot", rl.Vector2(x, y + 61), 75, 0, rl.WHITE)

    y += 160
    desc_font = gui_app.font(FontWeight.LIGHT)
    desc = "All processing is local. No cloud account required."
    wrapped = wrap_text(desc_font, desc, 45, int(w))
    for line in wrapped:
      rl.draw_text_ex(desc_font, line, rl.Vector2(x, y), 45, 0, rl.Color(200, 200, 200, 255))
      y += 45
